### PR TITLE
Migrate to text-conditioned prior training

### PR DESCRIPTION
**Pull request by nousr**
removed text embedding reader and opted to compute encodings and embeddings at train-time.

& some defaults changes pertaining to the switch, as well as making the default model larger (per Katherine's anecdotal evidence)

### CODE REVIEW

File: train_diffusion_prior.py [hunk 1]
```diff
@@ -154,10 +157,10 @@ def train(image_embed_dim,
...
 
     # Get image and text embeddings from the servers
-    print_ribbon("Downloading embeddings - image and text")
-    image_reader = EmbeddingReader(embeddings_folder=image_embed_url, file_format="npy")
-    text_reader  = EmbeddingReader(embeddings_folder=text_embed_url, file_format="npy")
-    num_data_points = text_reader.count
+    print_ribbon("Downloading Image Embeddings")
+    image_reader = EmbeddingReader(embeddings_folder=image_embed_url,
+                                   metadata_folder=meta_url, meta_columns=['caption'], file_format="parquet_npy")
+    num_data_points = image_reader.count
```
Comment (nousr) [inline train_diffusion_prior.py:163]: @rom1504 This is the biggest change, moving to a single EmbeddingReader. 

Given how much better Katherine's results are with text-conditioning I feel as though it is a safe bet to drop support for the embedding-only training scheme we had previously.

**Comment by rom1504** [comment]
I would advise strongly to keep the option to train with only precomputed text and image embeddings

It's good to add the option to use the text encodings but I think it's important to keep the precomputed only option

**Comment by nousr** [comment]
@rom1504 @lucidrains 

I made a wrapper that abstracts away the logic to handle both `embedding-only` and `text_conditioned+image_embedding`. 

Now it's a lot cleaner as you just call `for img, text in dataloader:` and it will automatically handle returning either the embeddings or tokenized text.

`make_splits()` will return a embedding wrapper for each of the requested splits (train/eval/test) which enables us to just pass around a single variable instead of `num_data_points, starts, stops` (etc).

**Comment by lucidrains** [comment]
@nousr ohhh, code looks 💯 💯  :)

**Comment by rom1504** [comment]
much nicer indeed!

does it work? can you share a wandb run?

**Comment by nousr** [comment]
> much nicer indeed!
> 
> does it work? can you share a wandb run?

@rom1504 
Let's find out! 🚀  Just launched these.
https://wandb.ai/nousr_laion/new-train-script?workspace=user-nousr_laion

File: train_diffusion_prior.py [hunk 2]
```diff
@@ -363,6 +363,7 @@ def train(
                 ### Evaluate model(validation run) ###
                 eval_model(diffusion_prior, eval_loader, dp_condition_on_text_encodings, dp_loss_type, phase="Validation")
 
+            step += 1
```
Comment (nousr) [inline train_diffusion_prior.py:366]: small bug here

File: dalle2_pytorch/dataloaders/embedding_wrapper.py [hunk 3]
```diff
@@ -20,7 +20,7 @@ def __init__(
         self.text_conditioned = text_conditioned
 
         if not self.text_conditioned:
-            self.text_reader = image_reader
+            self.text_reader = text_reader
```
Comment (nousr) [inline dalle2_pytorch/dataloaders/embedding_wrapper.py:23]: small bug here: fixed

Comment (nousr) [reply dalle2_pytorch/dataloaders/embedding_wrapper.py:23]: maybe a little bigger than small 😅 hahah

**Comment by nousr** [comment]
So @rom1504  the only thing I notice is that `NUM_TEST_EMBEDDINGS` is no longer used.

That means validation loops currently take much longer, should the default val-split just be very small? Or should I add a check inside the validation loop to break when we check more than specified.

**Comment by rom1504** [comment]
yes the validation scores computed often should be on a small part of the val set

**Comment by nousr** [comment]
> yes the validation scores computed often should be on a small part of the val set

alright @rom1504  I went ahead and just pushed some smaller values. this will at least let us prototype while we experiment with the conditioned prior. 

They're more or less arbitrary at this point (and ofc modifiable in the CLI)...I figure when we have some idea of how the prior trains we can go and fine-tune the values for val sets and the tracking intervals

**Comment by lucidrains** [comment]
lgtm!

**Pull request merged**